test/poaOperator.js

```diff
@@ -81,6 +99,71 @@ contract('PoaOperator', (accounts) => {
         assert.equal(p[2], proof[0]);
       });
     });
+
+    describe('CryptoEconomic Agregate Signatures', () => {
+      it('should allow to open and reject challenge', async () => {
+        const block = new Block(65);
+        const depositTx = Tx.deposit(0, 2000, alice);
+        block.addTx(depositTx);
+        const prevPeriodRoot = await bridge.tipHash();
+        const period = new Period(prevPeriodRoot, [block]);
+        period.setValidatorData(1, bob, CAS);
+        const proof = period.proof(depositTx);
+
+        await operator.submitPeriodWithCas(1, p[2], period.merkleRoot(), CAS, { from: bob }).should.be.fulfilled;
+        p[3] = await bridge.tipHash();
+        assert.equal(p[3], proof[0]);
+
+        const validatorRoot = merkelize(`0x000000000000000000000001${bob.replace('0x', '')}`, ZERO);
+        const consensusRoot = merkelize(period.merkleRoot(), ZERO);
+
+        await operator.challengeCas(CAS, validatorRoot, consensusRoot, 1, {value: '100000000000000000'});
+
+        let challenge = await operator.getChallenge(p[3], 1);
+        assert.equal(challenge[0], accounts[0]);
+        assert.equal(challenge[2], bob);
+
+        const casRoot = merkelize(CAS, validatorRoot);
+        const vote = Tx.periodVote(1, new Input(new Outpoint(consensusRoot, 0)));
+        vote.sign(['0x7bc8feb5e1ce2927480de19d8bc1dc6874678c016ae53a2eec6a6e9df717bfac']);
+        await operator.respondCas(consensusRoot, casRoot, 1, vote.inputs[0].v, vote.inputs[0].r, vote.inputs[0].s, accounts[0]);
```

Comment:
Two parties can call `respondCas`:
1. Slot owner itself
2. Someone else observing the chain

Slot owner has all the data in hand obviously.

Others can take consensus root and signature from the broadcasted period vote tx. Where are they supposed to take CAS bitmap from? Is it just the same submitted CAS with a challenged bit off ?

Reply:
it is the same CAS bitmap that was submitted with the period, no changes. We only need it here to check the bit, because it is not stored anywhere when period is submitted.

Reply:
ideally there is an endpoint somewhere in the node, where i can pull all data about a period submission:
- slotId
- signerAddr
- CAS bitmap
- blocksRoot
- prevPeriodRoot

this information is needed in some combination for all proofs.